### ISSUE TITLE
mkDerivation without an "out" output not possible

### ISSUE THREAD

**Issue by Profpatsch** [NixOS/nixpkgs#16182]
It appears nixpkgs makes it impossible to have no "out" output:

```
nix-build -E 'with import <nixpkgs> {}; stdenv.mkDerivation { name = "test"; outputs = [ "foo" ]; }'
these derivations will be built:
  /nix/store/23xfwynwyrqa8xnijf7faqhbq1x2laq5-test.drv
building path(s) ‘/nix/store/yjnv7jx3d7x832dxcx5lvdrqs9xd0ryb-test-foo’
Error: _assignFirst found no valid variant!
builder for ‘/nix/store/23xfwynwyrqa8xnijf7faqhbq1x2laq5-test.drv’ failed with exit code 1
```

cc @vcunat, based on the git blame of `_assignFirst`.

**Comment by vcunat** [comment]
Many output targets have "$out" as the default. If you don't provide that and don't provide an explicit override for them, you'll get this unfortunately unhelpful message.

**Comment by vcunat** [comment]
You can see the type groups and how to override them in http://hydra.nixos.org/job/nixpkgs/trunk/manual/latest/download/1/nixpkgs/manual.html#multiple-output-file-type-groups

**Comment by Profpatsch** [comment]
So I’d have to add `outputDev = "foo"` for it to work? But that gives the same error.

**Comment by cleverca22** [comment]
i think it needs to be set to $foo

**Comment by vcunat** [comment]
`foo` is correct but you apparently need to override multiple of them. In the linked docs there's a list with defaults, e.g. for `outputBin` the defaults are `bin` or `out`... so if you have neither you have to be explicit.

For a general fix, we should improve the error message. And/or we might work out some other changes to support `out`-less derivations better; I didn't think those would be common &ndash; note that e.g. `--prefix` is always set to `$out` (even if that's expanded to empty string) and similar things that go against...

**Comment by vcunat** [comment]
And, there's actually that suggestion to setup these variables during nix evaluation already and always export some aliases, e.g. `.dev` and `.bin`, even if they do *not* have separate outputs.

**Comment by vcunat** [comment]
I think it would be nice to have that before the next release, but there are way too many things I'd like to do...

**Comment by andrewrk** [comment]
> You can see the type groups and how to override them in

this link is broken now. can you put the content here?

**Comment by ledettwy** [comment]
ran into this today... I don't understand why out-less derivations would be too uncommon to be not worth supporting. That's going to be pretty much any non-FHS derivation that makes more than one logical "thing". It feels gross to have to keep a dummy $out around and not do anything with it.

**Comment by dzmitry-lahoda** [comment]
I get this. `Error: _assignFirst found no valid variant!` I read the doc. I do not understand how to use it.

**Comment by dzmitry-lahoda** [comment]
My case is genesis and keyring state for blockchain client server.

**Comment by Atemu** [comment]
I still don't get how you're supposed to make `out`-less derivations work. This should definitely be documented and/or fixed by defauly.

**Comment by vcunat** [comment]
TL;DR: it's not designed to allow `out`-less derivations (yet).

**Comment by Atemu** [comment]
A workaround is to have an empty `out` but have be at the second position or later in the `outputs` list because that list's order matters: The first output will be symlinked by nix-build etc. while all other ones are only accessible via the dedicated attributes.

If you want the bin output be the default (conventionally, you do), you need it to be at the first position in `outputs`.
